### PR TITLE
Improving ObjectSelector:

### PR DESCRIPTION
**Pull request by pefernan**
Merge after: https://github.com/droolsjbpm/droolsjbpm-build-bootstrap/pull/281

Improving ObjectSelector:
- Moving ListBox to TypeAhead widget

Improving ObjectSelector II:
- Improved Typeahead integration
- Added testing
- Fixing minor issues on Form Editor

### CODE REVIEW

**Comment by pefernan** [comment]
Jenkins retest this

File: kie-wb-common-forms/kie-wb-common-forms-commons/kie-wb-common-forms-common-rendering/kie-wb-common-forms-common-rendering-shared/src/test/java/org/kie/workbench/common/forms/commons/rendering/shared/util/masks/MaskInterpreterTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ */
+
+package org.kie.workbench.common.forms.commons.rendering.shared.util.masks;
+
+import junit.framework.TestCase;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public abstract class MaskInterpreterTest<T extends MaskInterpreter<Product>> extends TestCase {
```
Comment (csadilek) [inline kie-wb-common-forms/kie-wb-common-forms-commons/kie-wb-common-forms-common-rendering/kie-wb-common-forms-common-rendering-shared/src/test/java/org/kie/workbench/common/forms/commons/rendering/shared/util/masks/MaskInterpreterTest.java]: Shouldn't need to extend the old TestCase base class here.

File: kie-wb-common-forms/kie-wb-common-forms-commons/kie-wb-common-forms-common-rendering/kie-wb-common-forms-common-rendering-shared/src/test/java/org/kie/workbench/common/forms/commons/rendering/server/util/masks/BackendMaskInterpreterTest.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package org.kie.workbench.common.forms.commons.rendering.server.util.masks;
+
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.kie.workbench.common.forms.commons.rendering.shared.util.masks.MaskInterpreterTest;
+import org.kie.workbench.common.forms.commons.rendering.shared.util.masks.Product;
+
+@RunWith( JUnit4.class )
```
Comment (csadilek) [inline kie-wb-common-forms/kie-wb-common-forms-commons/kie-wb-common-forms-common-rendering/kie-wb-common-forms-common-rendering-shared/src/test/java/org/kie/workbench/common/forms/commons/rendering/server/util/masks/BackendMaskInterpreterTest.java]: shouldn't be needed. it should not longer be a junit3 runner used anywhere. if it is, we need to fix it.

File: kie-wb-common-forms/kie-wb-common-forms-commons/kie-wb-common-forms-common-rendering/kie-wb-common-forms-common-rendering-shared/src/main/java/org/kie/workbench/common/forms/commons/rendering/shared/util/masks/ModelInterpreter.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kie.workbench.common.forms.commons.rendering.shared.util.masks;
+
+public interface ModelInterpreter<T> {
```
Comment (csadilek) [inline kie-wb-common-forms/kie-wb-common-forms-commons/kie-wb-common-forms-common-rendering/kie-wb-common-forms-common-rendering-shared/src/main/java/org/kie/workbench/common/forms/commons/rendering/shared/util/masks/ModelInterpreter.java]: some javadoc would be nice

File: kie-wb-common-forms/kie-wb-common-forms-commons/kie-wb-common-forms-common-rendering/kie-wb-common-forms-common-rendering-shared/src/main/java/org/kie/workbench/common/forms/commons/rendering/shared/util/masks/MaskInterpreter.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+ */
+
+package org.kie.workbench.common.forms.commons.rendering.shared.util.masks;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.jboss.errai.common.client.api.Assert;
+
+public abstract class MaskInterpreter<T> {
```
Comment (csadilek) [inline kie-wb-common-forms/kie-wb-common-forms-commons/kie-wb-common-forms-common-rendering/kie-wb-common-forms-common-rendering-shared/src/main/java/org/kie/workbench/common/forms/commons/rendering/shared/util/masks/MaskInterpreter.java]: javadocs please.

**Comment by csadilek** [comment]
jenkins retest this.

**Comment by pefernan** [comment]
@csadilek changes done!

**Comment by csadilek** [comment]
jenkins retest this.

**Comment by csadilek** [comment]
jenkins retest this.

**Comment by csadilek** [comment]
jenkins retest this.

**Comment by csadilek** [comment]
build failure unrelated, @manstis do you want to take a look this PR as well?

**Comment by manstis** [comment]
@csadilek No, it's OK thank-you.. it should however be noted that the build failure means this PR is not actually being verified by Jenkins; as compilation of an upstream repository is failing (it also happens [here](https://github.com/droolsjbpm/kie-wb-common/pull/486). IMO we should not merge PRs that failed to be verified by our CI process - otherwise it's pretty pointless having the CI process.

**Comment by csadilek** [comment]
@manstis Agreed. I hadn't planned to merge it before it's green. Just thought you might want to take a look.

**Comment by csadilek** [comment]
jenkins retest this.

**Comment by pefernan** [comment]
jenkins retest this.

**Pull request merged**